keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,369 @@
+Volume Group KEP
+
+---
+title: Volume Group
+authors:
+  - "@xing-yang"
+  - "@jingxu97"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-storage
+reviewers:
+  - "@msau42"
+  - "@saad-ali"
+  - "@thockin"
+approvers:
+  - "@msau42"
+  - "@saad-ali"
+  - "@thockin"
+editor: TBD
+creation-date: 2020-02-12
+last-updated: 2020-02-12
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Title
+
+Volume Group
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [API Definitions](#api-definitions)
+  - [Example Yaml Files](#example-yaml-files)
+    - [Volume Group Snapshot](#volume-group-snapshot)
+    - [Volume Placement](#volume-placement)
+<!-- /toc -->
+
+## Summary
+
+This proposal is to introduce a VolumeGroup API to manage multiple volumes together and a GroupSnapshot API to take a snapshot of a VolumeGroup.
+
+## Motivation
+
+While there is already a KEP (https://github.com/kubernetes/enhancements/pull/1051) that introduces APIs to do application snapshot, backup, and restore, there are other use cases not covered by that KEP.
+
+Use case 1:
+A VolumeGroup allows users to manage multiple volumes belonging to the same application together and therefore it is very useful in general. For example, it can be used to group all volumes in the same StatefulSet together.
+
+User case 2:
+For some storage systems, volumes are always managed in a group. For these storage systems, they will have to create a group for a single volume if they need to implement a create volume function in Kubernetes. Providing a VolumeGroup API will be very convenient for them.
+
+Use case 3:
+Instead of taking individual snapshots one after another, VolumeGroup can be used as a source for taking a snapshot of all the volumes in the same volume group. This may be a storage level consistent group snapshot if the storage system supports it.  In any case, when used together with ExecutionHook, this group snapshot can be application consistent. For this use case, we will introduce another CRD GroupSnapshot.
```

Comment:
I don't see volumeGroup needs CSI changes, but, group snapshot may need.. am I correct with this understanding?

Reply:
volumeGroup also needs CSI changes. This depends on the storage system.  Some storage systems need to create a group ahead of time.  For others, this is just an entry in the k8s database, not on the storage system itself.

Reply:
@xing-yang iic, we have to replace the `ExecutionHook` with `containerNotifier` in new terms pov. Isnt it ?

Reply:
It's not finalized yet.  I'll make the changes if a decision is made.